src/types/mod.rs

```diff
@@ -27,17 +27,65 @@ pub use self::string::{PyString, PyString as PyUnicode};
 pub use self::tuple::PyTuple;
 pub use self::typeobject::PyType;
 
+// Implementations core to all native types
+#[macro_export]
+macro_rules! pyobject_native_type_base(
+    ($name: ty $(,generics: <$($generics:ident),+>)?) => {
+        unsafe impl<$($($generics),+)?> $crate::PyNativeType for $name {}
+
+        impl<$($($generics),+)?> ::std::fmt::Debug for $name {
+            fn fmt(&self, f: &mut ::std::fmt::Formatter)
+                   -> Result<(), ::std::fmt::Error>
+            {
+                let s = self.repr().map_err(|_| ::std::fmt::Error)?;
+                f.write_str(&s.to_string_lossy())
+            }
+        }
+
+        impl<$($($generics),+)?> ::std::fmt::Display for $name {
+            fn fmt(&self, f: &mut ::std::fmt::Formatter)
+                   -> Result<(), ::std::fmt::Error>
+            {
+                let s = self.str().map_err(|_| ::std::fmt::Error)?;
+                f.write_str(&s.to_string_lossy())
+            }
+        }
+
+        impl<$($($generics),+)?> $crate::ToPyObject for $name
+        {
+            #[inline]
+            fn to_object(&self, py: $crate::Python) -> $crate::PyObject {
+                use $crate::AsPyPointer;
+                unsafe { $crate::PyObject::from_borrowed_ptr(py, self.as_ptr()) }
+            }
+        }
+
+        impl<$($($generics),+)?> PartialEq for $name {
+            #[inline]
+            fn eq(&self, o: &$name) -> bool {
+                use $crate::AsPyPointer;
+
+                self.as_ptr() == o.as_ptr()
+            }
+        }
+    };
+);
+
+// Implementations core to all native types except for PyAny (because they don't
+// make sense on PyAny / have different implementations).
 #[macro_export]
 macro_rules! pyobject_native_type_named (
-    ($name: ty $(,$type_param: ident)*) => {
-        impl<$($type_param,)*> ::std::convert::AsRef<$crate::PyAny> for $name {
+    ($name: ty $(,generics: <$($generics:ident),+>)?) => {
+        $crate::pyobject_native_type_base!($name $(,generics: <$($generics),+>)?);
+
+        impl<$($($generics),+)?> ::std::convert::AsRef<$crate::PyAny> for $name {
             #[inline]
             fn as_ref(&self) -> &$crate::PyAny {
                 &self.0
             }
         }
 
-        impl<$($type_param,)*> ::std::ops::Deref for $name {
+        impl<$($($generics),+)?> ::std::ops::Deref for $name {
```

Comment:
I'm sorry but I don't understand why we need this complex one with an additional `$`.
Could you please explain this more?

Reply:
👍

So, first, for the exceptions macros, I wanted to make `$checkfunction` optional. They don't have their own checkfunctions, and so previously I had copy-pasted the default implementation of `PyTypeInfo::is_type_of`. By making `$checkfunction` optional I could get rid of this copy-and-paste.

However, then there were two optional `ident` arguments in a row: `$checkfunction` and `$type_param`. This is ambiguous for `macro_rules`, so it would not compile.

This meant I needed to change the syntax slightly. Because the generics are used least (i.e. only in rust-numpy as far as I know) I thought it was best to modify the syntax for the generics.

The solution here I made the additional optional form to pass to the macro `generics: <T, D>` . But because the `generics: < >` structure is all optional (`$(generics: <  ... >)?` this is what causes the extra `$` level.

I don't love it, but couldn't see many better options.

An alternative could be to change `$name:ty` to `$name:ident < $($generics),* >` and maybe add second macro arm to each for the generic-free case, but that's also kind of yucky....

I'm happy to accept any other proposals how to do this!